### PR TITLE
Improving axes grouping/splitting in static viz.

### LINKED ISSUE #20559: Static viz multi-series shows two Y-axis
**Describe the bug**
Static viz multi-series shows two Y-axis.
Kinda a regression in 0.42.0, since 0.41.6 would just show a table, but at least there wasn't a correctness issue with that.

**To Reproduce**
1. Question > Sample > Products
2. Summarize Sum of Price, grouped by CreatedAt:Year and Category
image
3. Visualize and save question
image
4. Not going to spend time on subscription, go to `/api/pulse/preview_card_png/:question_id`
image

**Expected behavior**
Follow Y-axis choices from normal visualization settings.

**Information about your Metabase Installation:**
Tested 0.41.6 and 0.42.0

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by adam-james-v**
fixes: #20559

### Before
We see that static viz renders with the axis split (we don't want to see the axis labels on the right):

image

### After
The axis labels are correctly being rendered only on the left:

image

Additionally:
- if a user explicitly sets any of the x-axis breakouts to display on the left or right, the setting is respected.
- axis grouping/splitting implementation works for both single-x-axis with multiple series and multi-x-axis.

### What's the Difference between Single-X-Axis and Multi-X-Axis?

This confused me at first, so I'll try to explain things. There are 2 ways in which a chart will have multiple lines (or areas or bars) rendered:

#### 1. A multi-x-axis situation.
This arises when you create a question with 2 (currently only 2) x-axes, with one of them having several breakouts. 

For example:
image

Which, in the data section as you set up your visualization, will look like this:
image

And, the shape of the rows that the query returns looks like this:

```clojure
;; X-axis      X-axis                Y-axis
[["Doohickey" "2016-01-01T00:00:00Z" 9031.557803943724]
 ["Doohickey" "2017-01-01T00:00:00Z" 43069.138089392625]
 ["Doohickey" "2018-01-01T00:00:00Z" 98515.46952597176]
 ["Doohickey" "2019-01-01T00:00:00Z" 110322.19082203327]
 ["Doohickey" "2020-01-01T00:00:00Z" 36332.63737703899]
 ["Gadget" "2016-01-01T00:00:00Z" 10672.625619920142]
 ["Gadget" "2017-01-01T00:00:00Z" 54960.620367033574]
 ["Gadget" "2018-01-01T00:00:00Z" 133811.6893419144]
 ["Gadget" "2019-01-01T00:00:00Z" 160502.6901169807]
 ["Gadget" "2020-01-01T00:00:00Z" 46673.62035739867]
 ["Gizmo" "2016-01-01T00:00:00Z" 9929.320704594422]
 ["Gizmo" "2017-01-01T00:00:00Z" 48130.70799398427]
 ["Gizmo" "2018-01-01T00:00:00Z" 136779.80350954592]
 ["Gizmo" "2019-01-01T00:00:00Z" 145044.8403051793]
 ["Gizmo" "2020-01-01T00:00:00Z" 43177.27364199722]
 ["Widget" "2016-01-01T00:00:00Z" 12523.365117960406]
 ["Widget" "2017-01-01T00:00:00Z" 59095.5552293035]
 ["Widget" "2018-01-01T00:00:00Z" 140938.0718003145]
 ["Widget" "2019-01-01T00:00:00Z" 161197.3266454905]
 ["Widget" "2020-01-01T00:00:00Z" 49913.1786806334]]
```

Where each row is one of the combinations of the X-Axis values, followed by the Y-Axis value for that combination. So, that first row is: _The sum of total orders of the product 'Doohickey' for the year 2016_.

It's worth recognizing that this is distinct from the next situation:

#### 2. Single-x-axis, with Multiple Series situation:
This was a bit easier for me to follow. This is when there is only 1 x-axis but any number of Y-Axes. For example (ignoring whether or not the question setup is sensible):

image

Those 3 'summarize' entries can be added to the Y-Axis section when setting up your visualization:
image

And the query returns the rows in the following shape:

```clojure
;; X-Axis value          Y-axis values
[["2022-04-03T00:00:00Z" 55 18 55]
 ["2022-04-04T00:00:00Z" 476 89 531]
 ["2022-04-05T00:00:00Z" 459 103 990]
 ["2022-04-06T00:00:00Z" 474 93 1464]
 ["2022-04-07T00:00:00Z" 455 99 1919]
 ["2022-04-08T00:00:00Z" 785 94 2704]
 ["2022-04-09T00:00:00Z" 87 26 2791]
 ["2022-04-10T00:00:00Z" 152 22 2943]
 ["2022-04-11T00:00:00Z" 548 92 3491]
 ["2022-04-12T00:00:00Z" 472 87 3963]
 ["2022-04-13T00:00:00Z" 641 116 4604]
 ["2022-04-14T00:00:00Z" 369 80 4973]
 ["2022-04-15T00:00:00Z" 265 56 5238]
 ["2022-04-16T00:00:00Z" 76 17 5314]
 ["2022-04-17T00:00:00Z" 6 4 5320]]
```

### CODE REVIEW

File: src/metabase/pulse/render/body.clj [hunk 1]
```diff
@@ -549,23 +549,53 @@
...
-      0)))
+        (/ (double overlap-width) (double max-width)))
+
+      ;; if either range is just a single point, and it's inside the other range,
+      ;; we consider it overlapped. Not likely in practice, but could happen.
+      (and (= min-a max-a) (<= min-b min-a max-b)) 1
+      (and (= min-b max-b) (<= min-a min-b max-a)) 1
+
+      ;; if we have a range that is a single point, is outside the other range, but
+      ;; is near it, we will calculate a 'nearness', which we will consider an overlap.
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:578]: Adding the concept of `nearness` sort of contradicts the name of the function being `overlap` which could be confusing. It's also a pretty long function so might be worth refactoring.

Here's a quick suggestion: split this out into two separate `overlap` and `nearness` functions and have them both called by a more generically-named function like `axis-group-threshold-val`.

File: src/metabase/pulse/render/body.clj [hunk 2]
```diff
@@ -642,41 +672,95 @@
...
+      {:unassigned? true :stacked? false}
+      (let [;; calculate the overlaps between every group.
+            ;; If there are 4 groups, there will be 4x4 overlaps calculated.
+            ;; TODO: the reciprocal pairings are technically unneeded calculation. It's optimizable,
+            ;; but the number of total groupings might never be high enough to really worry about it.
+            overlaps           (-> joined-rows-map
+                                   (update-vals (fn [vs]
+                                                  (into {} (map (fn [k]
+                                                                  [k (overlap (get joined-rows-map k) vs)])
+                                                                (keys joined-rows-map))))))
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:712]: Looks like the only use of `overlaps` is to get `relevant-distances`. Can't you just compute `relevant-distances` directly and skip creating the entire 2-d map of overlaps?

File: src/metabase/pulse/render/body.clj [hunk 3]
```diff
@@ -642,41 +672,95 @@
...
+            lefts              (or (:left positions) [(first (get positions nil))])
+            rights             (or (:right positions) [])
+            to-group           (remove (set (concat lefts rights)) (get positions nil))
+            relevant-distances (get overlaps (first lefts))
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:716]: nit: just call this `relevant-overlaps`? (or just `overlaps` if you address my comment above). probably best to use consistent terminology, and `distances` doesn't seem to be used elsewhere

File: src/metabase/pulse/render/body.clj [hunk 4]
```diff
@@ -642,41 +672,95 @@
...
-       :column        y-col})))
+  [chart-type joined-rows _x-cols y-cols {:keys [viz-settings] :as _data} card-name]
+  (let [positions (group-axes-at-once joined-rows viz-settings axis-group-threshold)]
+    (for [[idx y-col] (map-indexed vector y-cols)]
+      (let [y-col-key      (:name y-col)
+            card-type      (or (series-setting viz-settings y-col-key :display)
+                               chart-type
+                               (nth default-combo-chart-types idx))
+            selected-rows  (mapv #(vector (ffirst %) (nth (second %) idx)) joined-rows)
+            y-axis-pos     (or (get positions y-col-key) "left")]
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:740]: nit: `get` can take a default
```suggestion
            y-axis-pos     (get positions y-col-key "left")]
```

File: test/metabase/pulse/render/body_test.clj [hunk 5]
```diff
@@ -414,28 +415,28 @@
     (is (has-inline-image?
          (render-bar-graph {:cols         default-columns
                             :rows         [[10.0 1] [5.0 10] [2.50 20] [1.25 30]]
-                            :viz-settings {}}))))
+                            :viz-settings {:graph.metrics ["NumPurchased"]}}))))
```
Comment (noahmoss) [inline test/metabase/pulse/render/body_test.clj:418]: Curious why these all needed to be updated?

**Review by noahmoss — CHANGES_REQUESTED** [review]
Nice, this is complex stuff and looks solid overall. 👍 Requesting changes since I think I have a decent number of concrete asks in the comments but hopefully nothing is too onerous.

File: test/metabase/pulse/render/body_test.clj (hunk 5 above)
Comment (adam-james-v) [reply test/metabase/pulse/render/body_test.clj:418]: In a real question, if viz. has been set for line, area, or bar, the :viz-settings key will have :graph.metrics and :graph.dimensions values. 

The axis grouping strategy requires those keys to be able to actually re-construct the series. For example, a row `[timestamp 1 2 3 4 5]` needs to know that each of those values corresponds to a metric.

File: src/metabase/pulse/render/body.clj [hunk 6]
```diff
@@ -549,23 +549,53 @@
...
 
   overlap = (/ overlap-width max-width) = (/ 35 59) = 0.59"
-  [col-a col-b]
-  (let [[min-a min-b]    (map #(get-in % [:fingerprint :type :type/Number :min]) [col-a col-b])
-        [max-a max-b]    (map #(get-in % [:fingerprint :type :type/Number :max]) [col-a col-b])
-        valid-ranges?    (and min-a min-b max-a max-b
-                              ;; ranges with same min and max won't be considered ranges.
-                              (not= min-a max-a)
-                              (not= min-b max-b))
+  [vals-a vals-b]
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj:552]: Good catch! Changed the docstring

File: src/metabase/pulse/render/body.clj [hunk 7]
```diff
@@ -642,41 +672,95 @@
...
+        joined-rows-map    (if (= (count (ffirst joined-rows)) 2)
+                             ;; double-x-axis
+                             (-> (group-by (fn [[[_ x2] _]] x2) joined-rows)
+                                 (update-vals #(mapcat last %)))
+                             ;; single-x-axis
+                             (->> (:graph.metrics viz-settings)
+                                  (map-indexed (fn [idx k]
+                                                 [k (mapv #(get (second %) idx) joined-rows)]))
+                                  (into {})))
+        ;; map of group-key -> :left :right or nil
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj:671]: Yea, really good point. I used the key `:unassigned` as you suggested.

File: src/metabase/pulse/render/body.clj (hunk 4 above)
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj]: I always forget that. Thanks :)

File: src/metabase/pulse/render/body.clj [hunk 8]
```diff
@@ -642,41 +672,95 @@
...
+        ;; map of position (:left :right or nil) -> vector of 'assigned' groups
+        positions          (-> (group-by second starting-positions)
+                               (update-vals #(mapv first %)))
+        check              {:unassigned? (contains? positions nil)
+                            :stacked?    (boolean (:stackable.stack_type viz-settings))} ]
+    (case check
+      {:unassigned? true :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? false :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? true :stacked? false}
+      (let [;; calculate the overlaps between every group.
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj]: Yea, fair enough. I tried to clean things up and comment a bit more.

File: src/metabase/pulse/render/body.clj (hunk 3 above)
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj]: Good catch. I call things 'score' now, and I removed the creation of the whole map.

File: src/metabase/pulse/render/body.clj (hunk 2 above)
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj]: Yea, that's smart. Haha, thanks for pointing it out. I do that with `score-fn` instead now. Simpler!

File: src/metabase/pulse/render/body.clj [hunk 9]
```diff
@@ -642,41 +672,95 @@
...
+                                                 [k (mapv #(get (second %) idx) joined-rows)]))
+                                  (into {})))
+        ;; map of group-key -> :left :right or nil
+        starting-positions (into {} (for [k (keys joined-rows-map)]
+                                      [k (keyword (series-setting viz-settings k :axis))]))
+        ;; map of position (:left :right or nil) -> vector of 'assigned' groups
+        positions          (-> (group-by second starting-positions)
+                               (update-vals #(mapv first %)))
+        check              {:unassigned? (contains? positions nil)
+                            :stacked?    (boolean (:stackable.stack_type viz-settings))} ]
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/pulse/render/body.clj]: I realized I could use a cond instead for one fewer branches. Added comments to try clarify each cond as well.

File: src/metabase/pulse/render/body.clj [hunk 10]
```diff
@@ -549,50 +549,59 @@
...
-    (repeat "left")
-    (let [grouped-axes (-> (group-axes (:columns metadata) group-threshold)
-                           (update-vals count))]
-      (if (seq grouped-axes)
-        (mapcat (fn [k] (repeat (get grouped-axes k 1) (name k))) [:left :right])
-        (conj (repeat "right")
-              "left")))))
+  [vals-a vals-b]
+  (let [[min-a max-a]          (-> vals-a sort ((juxt first last)))
+        [min-b max-b]          (-> vals-b sort ((juxt first last)))
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:554]: nit: alignment

File: src/metabase/pulse/render/body.clj [hunk 11]
```diff
@@ -738,16 +801,16 @@
...
-        series-seqs     [(if (= (count x-cols) 1)
-                           (single-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name)
-                           (double-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name))]
-        labels          (combo-label-info x-cols y-cols viz-settings)
-        settings        (->ts-viz (first x-cols) (first y-cols) labels viz-settings)]
+        series-seqs   [(if (= (count x-cols) 1)
+                         (single-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name)
+                         (double-x-axis-combo-series enforced-type joined-rows x-cols y-cols data card-name))]
+        labels        (combo-label-info x-cols y-cols viz-settings)
+        settings      (->ts-viz (first x-cols) (first y-cols) labels viz-settings)]
```
Comment (noahmoss) [inline src/metabase/pulse/render/body.clj:813]: nit: I think the alignment before was correct here, no?

**Review by noahmoss — APPROVED** [review]

**Pull request merged**